molecule/provisioners/libvirtprovisioner.py

```diff
@@ -0,0 +1,923 @@
+        for index, inst in enumerate(self.molecule.config.config['libvirt'][
+                'instances']):
+            if inst['name'] == instance['name']:
+                if 'image' in self.molecule.config.config['libvirt'][
+                        'instances'][index]:
+                    self.molecule.config.config['libvirt']['instances'][index][
+                        'image'].update(image)
+                else:
+                    self.molecule.config.config['libvirt']['instances'][index][
+                        'image'] = image
+        return image
+
+    def up(self, no_provision=True):
+        """
+        Up or define/up libvirt instances.
+
+        :param: no_provision is not meaningful for libvirt instances
+        """
+        for net in self.molecule.config.config['libvirt']['networks']:
+            self._up_network(net)
+        pool = self._up_pool()
+        vols = pool.listAllVolumes()
+        domains = self._libvirt.listAllDomains()
+        wait = True
+        for instance in self.instances:
+            image = self._populate_image(instance)
+            imagefile = image['source'].split('/')[-1]
+            # Ensure that the image is available
+            if not os.path.exists(
+                    os.path.join(self._pool_path, image['name'] + '.img')):
+                if not os.path.exists(
+                        os.path.join(self._sources_path, imagefile)):
+                    self._fetch(image['source'], imagefile)
+                if imagefile.endswith('.box'):
+                    self._unpack_box(image, imagefile)
+
+            # Is there an existing libvirt volume for this instance?
+            try:
+                vol_found = pool.storageVolLookupByName(instance['name'])
+                if vol_found and not vol_found.isActive():
+                    vol_found.create()
+                return vol_found  # existing volume is now running
+            except libvirt.libvirtError as e:
+                self._create_volume(pool, instance)
+            # TODO: Are all this instance's interface's networks available?
+            # Is there an existing libvirt domain defined for this instance?
+            dom_found = False
+            for dom in domains:
+                if not dom_found:
+                    if dom.name() == instance['name']:
+                        dom_found = True
+                        if dom.info()[0] == 1:
+                            utilities.print_info(
+                                "\t{}: already running".format(instance[
+                                    'name']))
+                            break
+                        else:
+                            utilities.print_info("\t{}: booting".format(
+                                instance['name']))
+                            dom.create()
+                            if wait:
+                                time.sleep(self._boot_wait)
+                                wait = False
```

Comment:
If domain is running, then the rest if the up() method can be ignored. If the domain is found but not running (i.e. defined but not booted), then we want to boot and wait to allow it to boot completely. In either of those cases, we want to `continue` to the next instance.

If it's not defined, we need to define, boot and wait before moving on. I think this code does that, but the next commit makes this happen and the method is much clearer.